share/availability/light/availability.go

```diff
@@ -67,77 +66,88 @@ func (la *ShareAvailability) SharesAvailable(ctx context.Context, header *header
 		return nil
 	}
 
-	// do not sample over Root that has already been sampled
+	// load snapshot of the last sampling errors from disk
 	key := rootKey(dah)
-
 	la.dsLk.RLock()
-	exists, err := la.ds.Has(ctx, key)
+	last, err := la.ds.Get(ctx, key)
 	la.dsLk.RUnlock()
-	if err != nil || exists {
+
+	// Check for error cases
+	var samples []Sample
+	switch {
+	case err == nil && len(last) == 0:
+		// Availability has already been validated
+		return nil
+	case err != nil && !errors.Is(err, datastore.ErrNotFound):
+		// Other error occurred
 		return err
+	case errors.Is(err, datastore.ErrNotFound):
+		// No sampling result found, select new samples
+		samples, err = SampleSquare(len(dah.RowRoots), int(la.params.SampleAmount))
+		if err != nil {
+			return err
+		}
+	default:
+		// Sampling result found, unmarshal it
+		samples, err = decodeSamples(last)
+		if err != nil {
+			return err
+		}
 	}
 
-	log.Debugw("validate availability", "root", dah.String())
 	// We assume the caller of this method has already performed basic validation on the
 	// given dah/root. If for some reason this has not happened, the node should panic.
 	if err := dah.ValidateBasic(); err != nil {
 		log.Errorw("availability validation cannot be performed on a malformed DataAvailabilityHeader",
 			"err", err)
-		panic(err)
-	}
-	samples, err := SampleSquare(len(dah.RowRoots), int(la.params.SampleAmount))
-	if err != nil {
 		return err
 	}
 
 	// indicate to the share.Getter that a blockservice session should be created. This
 	// functionality is optional and must be supported by the used share.Getter.
 	ctx = getters.WithSession(ctx)
 
+	result := struct {
+		sync.Mutex
+		failed []Sample
+	}{}
```

Comment:
nit: this does not need to be a struct but two separate vars, which is more idiomatic